rainier-core/src/main/scala/com/stripe/rainier/sampler/LeapFrog.scala

```diff
@@ -60,6 +60,85 @@ private[sampler] case class LeapFrog(density: DensityFunction) {
     pqBuf(potentialIndex) = density.density * -1
   }
 
+  /**
+    * Perform l leapfrog steps starting at position q and momentum p
+    * @param l the total number of leapfrog steps to perform
+    * @return the new value of the parameters and momentum
+    * array with updated density
+    */
+  private def steps(l: Int, stepSize: Double): Unit = {
+    initialHalfThenFullStep(stepSize)
+    var i = 1
+    while (i < l) {
+      twoFullSteps(stepSize)
+      i += 1
+    }
+    finalHalfStep(stepSize)
+  }
+
+  private def isUTurn(theta: Array[Double], pqP: Array[Double]): Boolean = {
+
+    var out = 0.0
+    var i = 0
+    while (i < theta.size) {
+      out += (pqP(i + nVars) - theta(i)) * pqP(i)
+      i += 1
+    }
+
+    if (out.isNaN)
+      true
+    else
+      out < 0
+  }
+
+  /**
+    * Calculate the longest-step size until a u-turn
+    */
+  private def longestStepUntilUTurn(l0: Int, stepSize: Double): Int = {
+
+    val initTheta = variables(pqBuf)
+    var out = pqBuf
+    var l = 0
+    while (!isUTurn(initTheta, pqBuf)) {
+      l += 1
+      steps(1, stepSize)
+
+      if (l == l0)
+        out = pqBuf
+    }
+    copy(out, pqBuf)
+    l
+  }
+
+  /**
+    * Perform a single step of the longest batch step algorithm
+    */
+  private def longestBatchStep(l0: Int, stepSize: Double)(
+      implicit rng: RNG): Int = {
+
+    val params = pqBuf
+    initializePs(params)
+    val l = longestStepUntilUTurn(l0, stepSize)
+    if (l < l0)
+      steps(l0 - l, stepSize)
+    val u = rng.standardUniform
+    val a = logAcceptanceProb(params, pqBuf)
+    if (math.log(u) < a) {
+      copy(pqBuf, params)
+    }
+    l
+  }
+
+  /**
+    * Calculate a vector representing the empirical distribution
+    * of the steps taken until a u-turn
+    */
+  def empiricalLongestSteps(l0: Int, k: Int, stepSize: Double)(
+      implicit rng: RNG): Vector[Int] = {
+
+    Vector.fill(k)(longestBatchStep(l0, stepSize))
```

Comment:
Per the note in `EHMC`, I suspect you want to be passing in `params` here, and either passing it directly into `longestBatchStep`, or cloning it once and then using that clone for every call to `longestBatchStep`.